### PR TITLE
feat: added IAsyncEnumerable<T> query overloads to QueryAPI

### LINKED ISSUE #119: No timestamp returned via POCO based QueryAsync<T>
With a simple class like the following:
```C#
    [Measurement("point")]
    public class Point
    {
        [Column("tag", IsTag = true)] public string Tag { get; set; }
        [Column("min")] public double Minimum { get; set; }
        [Column("max")] public double Maximum { get; set; }
        [Column("avg")] public double Average { get; set; }
        [Column("flags")] public uint QualityFlags { get; set; }
        [Column(IsTimestamp = true)] public DateTime Timestamp { get; set; }
    }
```

Data writes fine and I can verify through a CLI query that timestamps are accurately stored, but when retriving data, the timestamp is always zero, i.e., `DateTime.MinValue`.

Here's the query code:

```C#
        public static async IAsyncEnumerable<Point> ReadPoints(string influxDBHost, string fluxQuery)
        {
            using InfluxDBClient influxDBClient = InfluxDBClientFactory.Create(influxDBHost, Token);
            QueryApi queryAPI = influxDBClient.GetQueryApi();

            List<Point> points = await queryAPI.QueryAsync<Point>($"from(bucket:\"{PointBucket}\"){fluxQuery}", OrganizationID);

            foreach (Point point in points)
                yield return point;
        }
```

Seems like the marked `IsTimestamp` column is being missed during deserialization.

Also, tried several variations on POCO timestamp column:

```C#
[Column("Timestamp", IsTimestamp = true)] public DateTime Timestamp;
[Column("time", IsTimestamp = true)] public DateTime Timestamp;
[Column("_time", IsTimestamp = true)] public DateTime Timestamp;
[Column(IsTimestamp = true)] public DateTime Time;
```

### PR DESCRIPTION
**Pull request by ritchiecarroll**
This adds [`IAsyncEnumerable<T>`](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8) support to the QueryAPI for .NET Standard 2.1 deployments. To support .NET Standard 2.1, client and client projects were modified to dual target .NET standard 2.1 and 2.0.

Closes #119 - both timestamp issue and suggested `IAsyncEnumerable` feature

### CODE REVIEW

File: Client.Core/Client.Core.csproj
Comment (bednar) [inline Client.Core/Client.Core.csproj:4]: 1. Please change order of versions and add `LangVersion`:
```suggestion
        <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
        <LangVersion>8</LangVersion>
```
2. Add `Microsoft.Bcl.AsyncInterfaces` ass conditional reference:
```xml
<ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
        <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
</ItemGroup>
```

File: Client/Client.csproj
Comment (bednar) [inline Client/Client.csproj:4]: Please change order of versions and add LangVersion:

```suggestion
      <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
      <LangVersion>8</LangVersion>  
```

**Review by bednar — CHANGES_REQUESTED** [review]
Hi @ritchiecarroll,

thanks for your PR!

Please add tests for your changes to be sure that everything works correctly.

Regards

**Comment by ritchiecarroll** [comment]
Requested changes and tests have been added to PR

**Comment by ritchiecarroll** [comment]
OK - I think I've done what I can here. Let me know how this looks.

File: Client.Core/Flux/Internal/FluxCsvParser.cs
Comment (bednar) [inline Client.Core/Flux/Internal/FluxCsvParser.cs:99]: Please remove this directive.

File: Client.Core/Internal/AbstractQueryClient.cs
Comment (bednar) [inline Client.Core/Internal/AbstractQueryClient.cs:108]: Please remove this directive.

File: Client.Legacy.Test/FluxCsvParserTest.cs
Comment (bednar) [inline Client.Legacy.Test/FluxCsvParserTest.cs:434]: Please remove this directive.

File: Client.Legacy.Test/FluxCsvParserTest.cs
Comment (bednar) [inline Client.Legacy.Test/FluxCsvParserTest.cs:486]: Please remove this directive.

File: Client.Core/Client.Core.csproj
Comment (bednar) [inline Client.Core/Client.Core.csproj:41]: Please add conditional dependency to `Microsoft.Bcl.AsyncInterfaces`:

```xml
<ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
        <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
</ItemGroup>
```

**Review by bednar — CHANGES_REQUESTED** [review]
Hi @ritchiecarroll,

there is a few things to change:

1. `Client.Legacy.csproj` should also be a dual targeting:
```
<TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
<LangVersion>8</LangVersion>
```

2. We could remove `#if NETSTANDARD2_1`and `#if NETCOREAPP3_1` by adding optional dependency to `Client.Core.csproj`:
```
<ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
     <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
</ItemGroup>
```

Could you give me a permission to update your `GridProtectionAlliance:feat/AsyncEnumerable` branch? I will fix our CircleCI scripts to supports dual targeting.

Regards

File: Client.Core/Client.Core.csproj
Comment (ritchiecarroll) [reply Client.Core/Client.Core.csproj:41]: > 
> Could you give me a permission to update your `GridProtectionAlliance:feat/AsyncEnumerable` branch? I will fix our CircleCI scripts to supports dual targeting.
> 
> Regards

Write permissions granted. Will attempt your other requests.

**Comment by ritchiecarroll** [comment]
> Hi @ritchiecarroll,
> 
> there is a few things to change:
> 
> 1. `Client.Legacy.csproj` should also be a dual targeting:
> 
> ```
> <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
> <LangVersion>8</LangVersion>
> ```
> 
> 1. We could remove `#if NETSTANDARD2_1`and `#if NETCOREAPP3_1` by adding optional dependency to `Client.Core.csproj`:
> 
> ```
> <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
>      <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
> </ItemGroup>
> ```
> 
> Could you give me a permission to update your `GridProtectionAlliance:feat/AsyncEnumerable` branch? I will fix our CircleCI scripts to supports dual targeting.
> 
> Regards

All requests completed. Also, check your e-mail associated with your GitHub account, you should have a pending invite for write permissions to the `GridProtectionAlliance:feat/AsyncEnumerable`  branch so that you can adjust CircleCI scripts as needed. I appreciate that step as I otherwise have to provide permissions to this tool for "all" repos I have write access to, which is more risk than I would like to incur until I trust this CI tool more.

Thanks,
Ritchie

**Comment by ritchiecarroll** [comment]
Hi @bednar - anything remaining for me to do here?

Thanks,
Ritchie

**Review by bednar — CHANGES_REQUESTED** [review]
Hi @ritchiecarroll, I've fixed CircleCI scripts to build your branch. 

Could you please rebase your PR with master and fix these failing tests: `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurements`, `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurementsWithoutFields`?
    
- https://app.circleci.com/pipelines/github/influxdata/influxdb-client-csharp/433/workflows/c658cc41-849b-46ca-8126-015b177fd560/jobs/1379

and again Thanks for you contribution.

**Comment by ritchiecarroll** [comment]
> Hi @ritchiecarroll, I've fixed CircleCI scripts to build your branch.
> 
> Could you please rebase your PR with master and fix these failing tests: `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurements`, `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurementsWithoutFields`?
> 
> * https://app.circleci.com/pipelines/github/influxdata/influxdb-client-csharp/433/workflows/c658cc41-849b-46ca-8126-015b177fd560/jobs/1379
> 
> and again Thanks for you contribution.

OK @bednar , fixed those.

FYI, I think the remaining failure on `dotnet-3.1-nightly` seems more related to this exception:
```
InfluxDB.Client.Core.Exceptions.HttpException : retention policy duration must be at least 1h0m0s
```
Is this a test configuration issue?

Thanks.
Ritchie

**Comment by bednar** [comment]
The remaining failure on `dotnet-3.1-nightly` seems to be related to nightly build of InfluxDB.

Please don't use a marge commits from master, because the PR contains unrelated changes. 

You should use rebase feature. This commands fix your git history:

```console
git remote add upstream https://github.com/influxdata/influxdb-client-csharp/
git fetch upstream

# reset before merge commit
git reset --hard 47e9727640976dcbaf0cb6f14af3c90bdac1d0eb
# rebase from master
git rebase upstream/master
# force push
git push -f origin feat/AsyncEnumerable
```

Thanks

**Comment by ritchiecarroll** [comment]
> The remaining failure on `dotnet-3.1-nightly` seems to be related to nightly build of InfluxDB.
> 
> Please don't use a marge commits from master, because the PR contains unrelated changes.
> 
OK @bednar, successful build after applying your specific rebase steps.

Thanks,
Ritchie

**Comment by ritchiecarroll** [comment]
@bednar: FYI, after successful build, the PR says "Merging can be performed automatically once the requested changes are addressed." but I don't think there is something I can do here, I believe this may be a step for you:

Changes requested
1 review requesting changes by reviewers with write access.

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (bednar) [inline Client.Core/Flux/Internal/FluxResultMapper.cs:158]: The following statement is unnecessary in `ToDateTimeValue`:

```csharp
if (value is DateTime dateTime)
{
      return dateTime;
}
```

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (bednar) [inline Client.Core/Flux/Internal/FluxResultMapper.cs:161]: The following statement is unnecessary in `ToInstantValue`:

```csharp
if (value is Instant instant)
{
     return instant;
}
```

**Review by bednar — CHANGES_REQUESTED** [review]
@ritchiecarroll we should also review an implementation, but we are on track to merge this to next release.

Thanks for your fantastic job here

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:158]: I do not feel this is true.  poco property may be `DateTime`, but object value coming may not be `DateTime`

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:161]: I do not feel this is true.  poco property may be `Instant`, but object value coming may not be `Instant`

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:158]: Or are you saying because of last line that already does something similar?

Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:158]: My thinking was this:

1) Handle case where value is of type `DateTime`
2) Handle case where value is of type `Instant`
3) Last chance, attempt cast to `DateTime` in case there are implict or explicit conversions available to convert to `DateTime` with associated value instance

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:161]: (see comment above for justification)

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:158]: OK, I updated implementation to fail appropriately when types do not match.

Comment (bednar) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:158]: Nice! Perfect improvement 👍

File: Client.Core/Flux/Internal/FluxCsvParser.cs
Comment (bednar) [inline Client.Core/Flux/Internal/FluxCsvParser.cs:8]: Using `System.Threading.Tasks` is not required by the code and can be safely removed

```suggestion
```

File: Client.Core/Flux/Internal/FluxCsvParser.cs
Comment (bednar) [inline Client.Core/Flux/Internal/FluxCsvParser.cs:80]: Use 'var' (simple types):

```suggestion
            var state = new ParseFluxResponseState { csv = csv };
```

File: Client.Core/Flux/Internal/FluxCsvParser.cs
Comment (bednar) [inline Client.Core/Flux/Internal/FluxCsvParser.cs:109]: Use 'var' (simple types):

```suggestion
            var state = new ParseFluxResponseState { csv = csv };
```

**Review by bednar — CHANGES_REQUESTED** [review]

File: Client.Core/Flux/Internal/FluxResultMapper.cs
Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:130]: I did a full update in code for converting specific type usage to `var` using ReSharper - there are many hundreds of cases that affects 251 files. I will update these per your request, if you would like me to replace all specific type instances with `var` in the entire project, may want to do that in a separate PR.

Comment (ritchiecarroll) [reply Client.Core/Flux/Internal/FluxResultMapper.cs:130]: Also, you can make a specific code styling normative by adding an [`.editorconfig`](https://docs.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options?view=vs-2019) file to your project.

See here for enforcing use of `var`:
https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-language-conventions?view=vs-2019#csharp_style_var_for_built_in_types

To help contributors understand the rules, here is an example [coding-style markdown](https://github.com/gemstone/common/blob/master/docs/coding-style.md) that can help verbally codify the pattern too.

Comment (bednar) [reply Client.Core/Flux/Internal/FluxResultMapper.cs]: Yeah, there are a lot of code style problems with autogenerated sources - `InfluxDB.Client.Api`... 😞 
Let's finish this PR and later we will see...

**Review by bednar — CHANGES_REQUESTED** [review]
@ritchiecarroll,

Could you please update `CHANGELOG.md`? After that we will be ready to merge it into master.

Thanks for your fantastic job! 💯

**Comment by ritchiecarroll** [comment]
Change log updated.

**Review by bednar — APPROVED** [review]

**Comment by ritchiecarroll** [comment]
@bednar just want to make sure you are not waiting on me to do something here. Thanks!

**Comment by bednar** [comment]
Hi @ritchiecarroll, the PR is merged into master. 

If you would like to use a preview version use our nightly build:

```
dotnet add package InfluxDB.Client --version 1.13.0-dev.1489
```

Regards

**Pull request merged**
